staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,238 @@ limitations under the License.
+	service.Name = name
+	info := &openAPISpecInfo{
+		apiService: service,
 	}
-	if len(specs) == 0 {
-		return &spec.Swagger{}, nil
-	}
-	sortByPriority(specs)
-	for _, specInfo := range specs {
-		if specToReturn == nil {
-			specToReturn = &spec.Swagger{}
-			*specToReturn = *specInfo.spec
-			// Paths and Definitions are set by MergeSpecsIgnorePathConflict
-			specToReturn.Paths = nil
-			specToReturn.Definitions = nil
-		}
-		if err := aggregator.MergeSpecsIgnorePathConflict(specToReturn, specInfo.spec); err != nil {
-			return nil, err
-		}
-	}
-
-	return specToReturn, nil
+	info.spec.Replace(spec)
+	s.openAPISpecs[name] = info
 }
 
-// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
-func (s *specAggregator) updateOpenAPISpec() error {
-	if s.openAPIVersionedService == nil {
-		return nil
-	}
-	specToServe, err := s.buildOpenAPISpec()
-	if err != nil {
-		return err
-	}
-	return s.openAPIVersionedService.UpdateSpec(specToServe)
-}
+// buildMergeSpecLocked creates a new cached mergeSpec from the list of cached.
+func (s *specAggregator) buildMergeSpecLocked() cached.Data[*spec.Swagger] {
+	apiServices := make([]v1.APIService, 0)
+	caches := make([]cached.Data[*spec.Swagger], len(s.openAPISpecs))
 
-// tryUpdatingServiceSpecs tries updating openAPISpecs map with specified specInfo, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
-	if specInfo == nil {
-		return fmt.Errorf("invalid input: specInfo must be non-nil")
+	for k := range s.openAPISpecs {
+		apiServices = append(apiServices, s.openAPISpecs[k].apiService)
 	}
-	_, updated := s.openAPISpecs[specInfo.apiService.Name]
-	origSpecInfo, existedBefore := s.openAPISpecs[specInfo.apiService.Name]
-	s.openAPISpecs[specInfo.apiService.Name] = specInfo
-
-	// Skip aggregation if OpenAPI spec didn't change
-	if existedBefore && origSpecInfo != nil && origSpecInfo.etag == specInfo.etag {
-		return nil
+	sortByPriority(apiServices)
+	for i, apiService := range apiServices {
+		caches[i] = &(s.openAPISpecs[apiService.Name].spec)
 	}
-	klog.V(2).Infof("Updating OpenAPI spec because %s is updated", specInfo.apiService.Name)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		reason := "add"
-		if updated {
-			reason = "update"
-		}
 
-		regenerationCounter.With(map[string]string{"apiservice": specInfo.apiService.Name, "reason": reason})
-		regenerationDurationGauge.With(map[string]string{"reason": reason}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		if existedBefore {
-			s.openAPISpecs[specInfo.apiService.Name] = origSpecInfo
-		} else {
-			delete(s.openAPISpecs, specInfo.apiService.Name)
+	return cached.NewListMerger(func(results []cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
```

Comment:
what is the semantics of this in the error case? Would this stop working when one aggregated apiserver fails? Would we keep going if there is an old cached copy?

Reply:
If an aggregated apiserver fails, the last successful copy will be returned. If there are no successes, it will be dropped and not aggregated. A single failing apiservice should not block the rest of the openapi from being aggregated.